### ISSUE TITLE
False negative: mypy ignores type error inside case that it deduces can't be reached

### ISSUE THREAD

**Issue by bkovitz** [python/mypy#14514]
**Bug Report**

Inside a `match` statement, _mypy_ 0.991 ignores type errors inside a `case` that it deduces can't be reached.

**To Reproduce**

```
from dataclasses import dataclass

@dataclass
class PaintAt:
    index: int

match 'a':
    case ('a',):   # <-- Doesn't match
        paint_at = PaintAt('wrong')  # <-- should raise type error
```

**Expected Behavior**

The line inside the `case` should fail to type-check.

In the example program above, the `case` really can't be reached. In my real code, though, the type of the `match` expression is more complicated, and is actually sometimes wrong at run-time. At run-time, the `case` does match the `match` expression. It would be nicer if `mypy` would catch type errors even inside cases that seem unreachable. This would help in using `mypy` to guide the programmer when making changes to code that temporarily break type correctness. The resulting error messages can point to exactly what needs to be changed next.

**Actual Behavior**

```
$ mypy --python-version 3.10 --show-error-codes so.py
Success: no issues found in 1 source file
```
**Your Environment**

- Mypy version used: 0.991
- Mypy command-line flags:  --python-version 3.10 --show-error-codes
- Mypy configuration options from `mypy.ini` (and other config files): (none)
- Python version used: 3.10.9
- Operating system and version: Mac OSX 12.4

**Comment by hauntsaninja** [comment]
mypy complains about this with `--warn-unreachable`
